executor/sort.go

```diff
@@ -190,7 +220,72 @@ func (e *SortExec) prepareExternalSorting() (err error) {
 	e.rowChunks = nil
 	e.partitionList = append(e.partitionList, listInDisk)
 	e.partitionRowPtrs = append(e.partitionRowPtrs, e.initPointersForListInDisk(listInDisk))
-	return err
+	return nil
+}
+
+type topNChunkHeapWithIndex struct {
+	*SortExec
+}
+
+func (h *topNChunkHeapWithIndex) Less(i, j int) bool {
+	rowI := h.sortRows[i]
+	rowJ := h.sortRows[j]
+	return h.lessRow(rowI, rowJ)
+}
+
+func (h *topNChunkHeapWithIndex) Len() int {
+	return len(h.sortRows)
+}
+
+func (h *topNChunkHeapWithIndex) Push(x interface{}) {
+	// Should never be called.
+}
+
+func (h *topNChunkHeapWithIndex) Pop() interface{} {
+	return nil
+}
+
+func (h *topNChunkHeapWithIndex) Swap(i, j int) {
+	h.sortRows[i], h.sortRows[j] = h.sortRows[j], h.sortRows[i]
+	h.sortRowsIndex[i], h.sortRowsIndex[j] = h.sortRowsIndex[j], h.sortRowsIndex[i]
+}
+
+func (e *SortExec) externalSorting(req *chunk.Chunk) (err error) {
+	if e.heapSort == nil {
+		e.heapSort = &topNChunkHeapWithIndex{e}
+		heap.Init(e.heapSort)
+		for i := 0; i < len(e.partitionList); i++ {
+			e.partitionConsumedRows[i] = 0
+			row, err := e.partitionList[i].GetRow(e.partitionRowPtrs[i][0])
+			if err != nil {
+				return err
+			}
+			e.sortRows = append(e.sortRows, row)
+			e.sortRowsIndex = append(e.sortRowsIndex, i)
+			heap.Fix(e.heapSort, 0)
+		}
+	}
+
+	for !req.IsFull() && e.heapSort.Len() > 0 {
+		length := e.heapSort.Len() - 1
+		heap.Pop(e.heapSort)
```

Comment:
We should not call `heap.Pop` since you want to change the least and use `heap.Fix` later.